router/internal/graphqlmetrics/batch_queue.go

```diff
@@ -0,0 +1,163 @@
+package graphqlmetrics
+
+import "time"
+
+// Inspired by https://github.com/wind-c/bqueue
+
+import (
+	"context"
+	"sync/atomic"
+)
+
+const (
+	defaultInterval      = time.Duration(5) * time.Second
+	defaultMaxBatchItems = 1024
+	defaultMaxQueueSize  = 1024
+)
+
+// BatchQueueOptions configure time interval and set a batch limit.
+type BatchQueueOptions struct {
+	Interval      time.Duration // wait time when batch quantity is insufficient
+	MaxBatchItems int           // maximum items per batch
+	MaxQueueSize  int           // maximum queue size
+}
+
+func (o *BatchQueueOptions) ensureDefaults() {
+	if o.Interval == 0 {
+		o.Interval = defaultInterval
+	}
+
+	if o.MaxBatchItems == 0 {
+		o.MaxQueueSize = defaultMaxBatchItems
+	}
+
+	if o.MaxQueueSize == 0 {
+		o.MaxQueueSize = defaultMaxQueueSize
+	}
+}
+
+// BatchQueue coordinates dispatching of queue items by time intervals
+// or immediately after the batching limit is met.
+type BatchQueue struct {
+	config          *BatchQueueOptions
+	ctx             context.Context
+	cancel          context.CancelFunc
+	doWork          chan struct{}
+	stopped         *atomic.Bool
+	timer           *time.Timer
+	inQueue         chan any
+	midQueue        chan any
+	OutQueue        chan []any
+	dispatchedCount int
+}
+
+// NewBatchQueue returns an initialized instance of BatchQueue.
+// Items are enqueued without blocking.
+// The queue is dispatched by time intervals or immediately after the batching limit is met.
+// Batches can be read from the OutQueue channel.
+func NewBatchQueue(config *BatchQueueOptions) *BatchQueue {
+	if config == nil {
+		config = new(BatchQueueOptions)
+	}
+	config.ensureDefaults()
+
+	bq := &BatchQueue{
+		config:          config,
+		doWork:          make(chan struct{}),
+		inQueue:         make(chan any, config.MaxQueueSize/2),
+		midQueue:        make(chan any, config.MaxQueueSize/2),
+		OutQueue:        make(chan []any, config.MaxQueueSize/config.MaxBatchItems),
+		stopped:         &atomic.Bool{},
+		dispatchedCount: 0,
+	}
+
+	return bq
+}
+
+// Enqueue adds an item to the queue. Returns false if the queue is stopped or not ready to accept items
+func (b *BatchQueue) Enqueue(item any) bool {
+	if b.stopped.Load() {
```

Comment:
You're currently not evaluating the result of Enqueue in the http handler.
Is it really necessary to have this atomic and load it?
If the server behaviour will not adopt to it, could you not just omit it?

Reply:
We can omit it.

Reply:
Done